[FLINK-20188][Connectors][Docs][FileSystem] Added documentation for File Source (#18288)
apache/flink: docs/content/docs/connectors/datastream/filesystem.md

under the License.
-->

# FileSystem

This connector provides a unified Source and Sink for `BATCH` and `STREAMING` that reads or writes (partitioned) files to filesystems
supported by the [Flink `FileSystem` abstraction]({{< ref "docs/deployment/filesystems/overview" >}}). This filesystem
connector provides the same guarantees for both `BATCH` and `STREAMING` and is designed for providing exactly-once semantics for `STREAMING` execution.

The connector supports reading and writing a set of files from any (distributed) file system (e.g. POSIX, S3, HDFS)
with a [format]({{< ref "docs/connectors/datastream/formats/overview" >}}) (e.g., Avro, CSV, Parquet),
producing a stream or records.

## File Source

The `File Source` is based on the [Source API]({{< ref "docs/dev/datastream/sources" >}}#the-data-source-api), 
a unified data source that reads files - both in batch and in streaming mode. 

> It is divided into the following two parts: File SplitEnumerator and File SourceReader. 

Reviewer: Since the similar part for internal components has been deleted, should we also delete this part?

Reviewer: True. There is a section about "Customizing File Enumeration" below. I am afraid that user might ask question about the relationship between the FileEnumerator and SplitEnumerator.

Author: That's a good point. Perhaps that should actually be removed.

You'll need to combine the File Source with a [format]({{< ref "docs/connectors/datastream/formats/overview" >}}). This allows you to
parse CSV, decode AVRO or read Parquet columnar files.

#### Bounded and Unbounded Streams

A bounded `File Source` lists all files (via SplitEnumerator, for example a recursive directory list with filtered-out hidden files) and reads them all.

An unbounded `File Source` is created when configuring the enumerator for periodic file discovery.
In that case, the SplitEnumerator will enumerate like the bounded case but after a certain interval repeats the enumeration.
For any repeated enumeration, the `SplitEnumerator` filters out previously detect files and only sends new ones to the `SourceReader`.

### Usage

You start building a file source via one of the following calls:

{{< tabs "FileSourceUsage" >}}
{{< tab "Java" >}}
```java
// reads the contents of a file from a file stream. 
FileSource.forRecordStreamFormat(StreamFormat,Path...)
        
// reads batches of records from a file at a time
FileSource.forBulkFileFormat(BulkFormat,Path...)
```
{{< /tab >}}
{{< /tabs >}}

> This creates a `FileSource.FileSourceBuilder` on which you can configure all the properties of the file source.
> 
> For the bounded/batch case, the file source processes all files under the given path(s). 
> In the continuous/streaming case, the source periodically checks the paths for new files and will start reading those.
> 
> When you start creating a file source (via the `FileSource.FileSourceBuilder` created through one of the above-mentioned methods) 
> the source is by default in bounded/batch mode. Call `AbstractFileSource.AbstractFileSourceBuilder.monitorContinuously(Duration)` 
> to put the source into continuous streaming mode.

Reviewer 2: ```suggestion
This creates a `FileSource.FileSourceBuilder` on which you can configure all the properties of the File Source.

For the bounded/batch case, the File Source processes all files under the given path(s). 
For the continuous/streaming case, the source periodically checks the paths for new files and will start reading those.

When you start creating a File Source (via the `FileSource.FileSourceBuilder` created through one of the above-mentioned methods), 
the source is in bounded/batch mode by default. You can call `AbstractFileSource.AbstractFileSourceBuilder.monitorContinuously(Duration)` 
to put the source into continuous streaming mode.
```

{{< tabs "FileSourceBuilder" >}}
{{< tab "Java" >}}
```java
final FileSource<String> source =
        FileSource.forRecordStreamFormat(...)
        .monitorContinuously(Duration.ofMillis(5))  
        .build();
```
{{< /tab >}}
{{< /tabs >}}

### Format Types

The reading of each file happens through file readers defined by file formats. 
These define the parsing logic for the contents of the file. There are multiple classes that the source supports. 

> Their interfaces trade of simplicity of implementation and flexibility/efficiency.

Reviewer 2: ```suggestion
The interfaces are a tradeoff between simplicity of implementation and flexibility/efficiency.
```

* A `StreamFormat` reads the contents of a file from a file stream. It is the simplest format to implement, 
and provides many features out-of-the-box (like checkpointing logic) but is limited in the optimizations it can apply 
(such as object reuse, batching, etc.).

* A `BulkFormat` reads batches of records from a file at a time. 
It is the most "low level" format to implement, but offers the greatest flexibility to optimize the implementation.

> #### TextLine format
> 
> A `StreamFormat` reader format that text lines from a file.
> The reader uses Java's built-in `InputStreamReader` to decode the byte stream using
> various supported charset encodings.
> This format does not support optimized recovery from checkpoints. On recovery, it will re-read
> and discard the number of lined that were processed before the last checkpoint. That is due to
> the fact that the offsets of lines in the file cannot be tracked through the charset decoders
> with their internal buffering of stream input and charset decoder state.

Reviewer 2: ```suggestion
#### TextLine Format

A `StreamFormat` reader formats text lines from a file.
The reader uses Java's built-in `InputStreamReader` to decode the byte stream using
various supported charset encodings.
This format does not support optimized recovery from checkpoints. On recovery, it will re-read
and discard the number of lines that were processed before the last checkpoint. This is due to
the fact that the offsets of lines in the file cannot be tracked through the charset decoders
with their internal buffering of stream input and charset decoder state.
```

#### SimpleStreamFormat Abstract Class

> A simple version of `StreamFormat` for formats that are not splittable.

Reviewer 2: ```suggestion
This is a simple version of `StreamFormat` for formats that are not splittable.
```

Custom reads of Array or File can be done by implementing `SimpleStreamFormat`:

{{< tabs "SimpleStreamFormat" >}}
{{< tab "Java" >}}
```java
private static final class ArrayReaderFormat extends SimpleStreamFormat<byte[]> {
    private static final long serialVersionUID = 1L;

    @Override
    public Reader<byte[]> createReader(Configuration config, FSDataInputStream stream)
            throws IOException {
        return new ArrayReader(stream);
    }

    @Override
    public TypeInformation<byte[]> getProducedType() {
        return PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO;
    }
}

final FileSource<byte[]> source =
                FileSource.forRecordStreamFormat(new ArrayReaderFormat(), path).build();
```
{{< /tab >}}
{{< /tabs >}}

> An example of a `SimpleStreamFormat` is `CsvReaderFormat`. It can be initialized as follows:

Reviewer 2: ```suggestion
An example of a `SimpleStreamFormat` is `CsvReaderFormat`. It can be initialized like this:
```

    
    // reference constructor
    public HiveSourceFileEnumerator(...) {
        ...
    }

    /***
     * Generates all file splits for the relevant files under the given paths. The {@code
     * minDesiredSplits} is an optional hint indicating how many splits would be necessary to
     * exploit parallelism properly.
     */
    @Override
    public Collection<FileSourceSplit> enumerateSplits(Path[] paths, int minDesiredSplits)
            throws IOException {
        // createInputSplits:splitting files into fragmented collections
        return new ArrayList<>(createInputSplits(...));
    }

    ...

    /***
     * A factory to create HiveSourceFileEnumerator.
     */
    public static class Provider implements FileEnumerator.Provider {

        ...
        @Override
        public FileEnumerator create() {
            return new HiveSourceFileEnumerator(...);
        }
    }
}
// use the customizing file enumeration
new HiveSource<>(
        ...,
        new HiveSourceFileEnumerator.Provider(
        partitions != null ? partitions : Collections.emptyList(),
        new JobConfWrapper(jobConf)),
       ...);
```
{{< /tab >}}
{{< /tabs >}}

### Current Limitations

> Watermarking doesn't work particularly well for large backlogs of files, because watermarks eagerly advance within a file, and the next file might contain data later than the watermark again.
> 

Reviewer 2: ```suggestion
Watermarking does not work very well for large backlogs of files. This is because watermarks eagerly advance within a file, and the next file might contain data later than the watermark.

```

> For Unbounded File Sources, the enumerator currently remembers paths of all already processed files, which is a state that can in some cases grow rather large.
> The future will be planned to add a compressed form of tracking already processed files in the future (for example by keeping modification timestamps lower boundaries).

Reviewer 2: ```suggestion
For Unbounded File Sources, the enumerator currently remembers paths of all already processed files, which is a state that can, in some cases, grow rather large.
There are plans to add a compressed form of tracking already processed files in the future (for example, by keeping modification timestamps below boundaries).
```

> ### Behind the Scene
> {{< hint info >}}
> If you are interested in how File source works under the design of new data source API, you may
> want to read this part as a reference. For details about the new data source API,
> [documentation of data source]({{< ref "docs/dev/datastream/sources.md" >}}) and
> <a href="https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface">FLIP-27</a>
> provide more descriptive discussions.
> {{< /hint >}}

Reviewer 2: ```suggestion
### Behind the Scenes
{{< hint info >}}
If you are interested in how File Source works through the new data source API design, you may
want to read this part as a reference. For details about the new data source API, check out the
[documentation on data sources]({{< ref "docs/dev/datastream/sources.md" >}}) and
<a href="https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface">FLIP-27</a>
for more descriptive discussions.
{{< /hint >}}
```

     * minDesiredSplits} is an optional hint indicating how many splits would be necessary to
     * exploit parallelism properly.
     */
    @Override
    public Collection<FileSourceSplit> enumerateSplits(Path[] paths, int minDesiredSplits)
            throws IOException {
        // createInputSplits:splitting files into fragmented collections
        return new ArrayList<>(createInputSplits(...));
    }

    ...

    /***
     * A factory to create HiveSourceFileEnumerator.
     */
    public static class Provider implements FileEnumerator.Provider {

        ...
        @Override
        public FileEnumerator create() {
            return new HiveSourceFileEnumerator(...);
        }
    }
}
// use the customizing file enumeration
new HiveSource<>(
        ...,
        new HiveSourceFileEnumerator.Provider(
        partitions != null ? partitions : Collections.emptyList(),
        new JobConfWrapper(jobConf)),
       ...);
```
{{< /tab >}}
{{< /tabs >}}

### Current Limitations

Watermarking doesn't work particularly well for large backlogs of files, because watermarks eagerly advance within a file, and the next file might contain data later than the watermark again.

> We are looking at ways to generate the watermarks more based on global information.

Reviewer 3: I cannot recall seeing any planned effort to fix this behavior.

Author: This was a copy & paste from Stephan's input. I'll remove it

> For Unbounded File Sources, the enumerator currently remembers paths of all already processed files, which is a state that can in come cases grow rather large.

Reviewer 3: ```suggestion
For Unbounded File Sources, the enumerator currently remembers paths of all already processed files, which is a state that can in some cases grow rather large.
```

Author: Thanks, fixed

> The future will be planned to add a compressed form of tracking already processed files in the future (for example by keeping modification timestamps lower boundaries).

Reviewer 3: The delta lake source also hit this problem we probably have to fix this rather soonish.

Author: Do you want to mention that in the docs (with a link to a Jira ticket)?

Author: Created https://issues.apache.org/jira/browse/FLINK-25672 for this

### Behind the Scene
{{< hint info >}}
If you are interested in how File source works under the design of new data source API, you may
want to read this part as a reference. For details about the new data source API,
[documentation of data source]({{< ref "docs/dev/datastream/sources.md" >}}) and
<a href="https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface">FLIP-27</a>
provide more descriptive discussions.
{{< /hint >}}

> The `File Source` is divided in the following two parts: File SplitEnumerator and File SourceReader.

Reviewer 3: I'd remove everything beginning from here because the internals might be outdated or wrong and on the other not confusing for users.

Author: Done

The Source of a file system is divided into the following two parts: File Split Enumerator and File Source Reader,
File Split Enumerator (Split is an abstraction of the external file system data splitting)
It is responsible for discovering the split in the external system and assigning it to the File SourceReader,

> and it also manages the global water level to ensure that the consumption rate is approximately the same between our different File Source Readers.

Reviewer 3: I think this sentence is partially a lie because afaik watermark alignment is not supported by any source yet.

Author: Removed it.
